spec.md

```diff
@@ -0,0 +1,954 @@
+}
+```
+
+##### GetPluginInfo Errors
+
+If the plugin is unable to complete the GetPluginInfo call successfully, it MUST return a non-ok gRPC code in the gRPC status.
+
+#### `GetPluginCapabilities`
+
+This REQUIRED RPC allows the CMI-Client to query the supported capabilities of the Plugin "as a whole": it is the grand sum of all capabilities.
+All CMI-Plugins of the same version (see `version` of `GetPluginInfoResponse`) of the Plugin SHALL return the same set of capabilities.
+
+```protobuf
+
+message GetPluginCapabilitiesRequest {
+    // Intentionally empty.
+}
+
+message GetPluginCapabilitiesResponse {
+    // All the capabilities that the controller service supports.
+    // This field is OPTIONAL.
+    repeated PluginCapability capabilities = 1;
+}
+
+// Specifies a capability of the CMI-Plugin.
+message PluginCapability {
+    message RPC {
+        enum Type {
+            // UNKNOWN is used to specific an capaability beyond the set
+            // provided below
+            UNKNOWN = 0;
+
+            // CREATE_MACHINE tells that the plugin implements the
+            // CreateMachine() interface
+            CREATE_MACHINE = 1;
+
+            // DELETE_MACHINE tells that the plugin implements the
+            // DeleteMachine() interface
+            DELETE_MACHINE = 2;
+
+            // GET_MACHINE tells that the plugin implements the
+            // GetMachine() interface
+            GET_MACHINE = 3;
+
+            // SHUTDOWN_MACHINE tells that he plugin implements the
+            // ShutDownMachine() interface
+            SHUTDOWN_MACHINE = 4;
+
+            // GET_LIST_OF_VOLUMEIDS_FOR_EXISTING_PVS tells if the plugin
+            // implements the GetListOfVolumeIDsForExistingPVs() interface
+            GET_LIST_OF_VOLUMEIDS_FOR_EXISTING_PVS = 5;
+        }
+        Type type = 1;
+    }
+    oneof type {
+        // RPC that the plugin supports.
+        RPC rpc = 1;
+    }
+}
+```
+
+##### GetPluginCapabilities Errors
+
+If the plugin is unable to complete the GetPluginCapabilities call successfully, it MUST return a non-ok gRPC code in the gRPC status.
+
+#### `Probe`
+
+A Plugin MUST implement this RPC call.
+The primary utility of the Probe RPC is to verify that the plugin is in a healthy and ready state.
+If an unhealthy state is reported, via a non-success response, a CMI-Client MAY take action with the intent to bring the plugin to a healthy state.
+Such actions MAY include, but SHALL NOT be limited to, the following:
+
+* Restarting the plugin container, or
+* Notifying the plugin supervisor.
+
+The Plugin MAY verify that it has the right configurations, devices, dependencies and drivers in order to run and return a success if the validation succeeds.
+The CMI-Client MAY invoke this RPC at any time.
+A CMI-Client MAY invoke this call multiple times with the understanding that a plugin's implementation MAY NOT be trivial and there MAY be overhead incurred by such repeated calls.
```

Comment:
`A CMI-Client MAY invoke this call multiple times with the understanding that a plugin's implementation MAY NOT be trivial and there MAY be overhead incurred by such repeated calls.`

Would it make sense to introduce a sleep for such repeated calls to the plugin container, similar behaviour was introduced in Etcd-backup-restore recently as a similar liveness check was running too often and was sub-optimal. wdyt ?

Reply:
The implementation of the `CMI-Client` would definitely make these calls after regular intervals (we are even thinking of exponential). However, this is just a heads up to plugin implementor to be ready for simultaneous/multiple calls.

Reply:
I have added a section on the retry mechanism. https://github.com/gardener/machine-spec/pull/11/files#diff-958e7270f96f5407d7d980f500805b1bR205-R211